File: ios/Cloudmoji/Cloudmoji/Views/ParentalGate.swift
import SwiftUI

/// One arithmetic question.
///
/// Ported from `src/components/ParentalGate.tsx`, including the eight pairs and the
/// rotation. Rotating rather than randomising is deliberate: the gate does not need
/// unpredictability — a two-year-old cannot do arithmetic at all — and a fixed
/// sequence is testable.
struct GateChallenge: Equatable {
    let a: Int
    let b: Int

    var answer: Int { a * b }

    static let all: [GateChallenge] = [
        GateChallenge(a: 7, b: 8),
        GateChallenge(a: 9, b: 6),
        GateChallenge(a: 12, b: 7),
        GateChallenge(a: 6, b: 11),
        GateChallenge(a: 8, b: 9),
        GateChallenge(a: 11, b: 8),
        GateChallenge(a: 7, b: 12),
        GateChallenge(a: 9, b: 7),
    ]

    /// Wraps, so a parent opening Settings a ninth time gets the first question
    /// again rather than an index out of range.
    static func at(_ index: Int) -> GateChallenge {
        all[abs(index) % all.count]
    }

    /// The whole gate, in one line.
    ///
    /// Trimming is not leniency: a numeric keypad and an autocorrect pass both
    /// produce trailing whitespace, and a parent who typed the right answer has
    /// answered the question. Anything that is not a whole number is not an answer
    /// — which is already stricter than the web, where `Number("")` is 0.
    func accepts(_ entry: String) -> Bool {
        guard let value = Int(entry.trimmingCharacters(in: .whitespaces)) else { return false }
        return value == answer
    }
}

/// A real gate, not a gesture.
///
/// Deliberately boring: no timer, no penalty, no lock-out. A parent who misreads
/// just tries again. Ported from `src/components/ParentalGate.tsx`.
///
/// Drawn as an overlay rather than a sheet, for two reasons that both bit: a sheet
/// swapped for another sheet on pass is a documented SwiftUI misfire, and a numeric
/// keyboard inside a detented sheet fights the detent.
struct ParentalGate: View {
    let challenge: GateChallenge
    /// What the parent is about to do, so they know why they were asked.
    let action: String
    let onPass: () -> Void
    let onCancel: () -> Void

    @State private var entry = ""
    @State private var wasWrong = false
    @FocusState private var isFocused: Bool

    private static let cardWidth: CGFloat = 340
    /// Parent-only chrome, so the 44pt HIG minimum. The web draws 52 and 48; both
    /// clear it and both are kept.
    private static let fieldHeight: CGFloat = 52
    private static let buttonHeight: CGFloat = 48

    var body: some View {
        ZStack {
            Color.black.opacity(0.82)
                .ignoresSafeArea()
                // Tapping outside is Cancel. A toddler who reached this screen
                // needs a way out that is not the right answer.
                .onTapGesture(perform: onCancel)

            card
                .frame(maxWidth: Self.cardWidth)
                .padding(20)
        }
        // Without this the identifier below propagates down and overwrites
        // `gate-input`, `gate-submit`, `gate-cancel` and `gate-question` — the
        // same defect that made all three of the typing row's controls
        // unreachable for the whole of stage 2a. It is invisible to every unit
        // test, because SwiftUI builds no accessibility tree outside XCUITest.
        .accessibilityElement(children: .contain)
        .accessibilityIdentifier("parental-gate")
        .task { isFocused = true }
    }

    private var card: some View {
        VStack(alignment: .leading, spacing: 0) {
            Text("Grown-ups only")
                .font(Theme.body(16, .black))
                .foregroundStyle(Theme.textPrimary)
                .padding(.bottom, 6)

            Text(action)
                .font(Theme.body(13, .bold))
                .foregroundStyle(Theme.textSecondary)
                .fixedSize(horizontal: false, vertical: true)
                .padding(.bottom, 16)

            Text("What is \(challenge.a) × \(challenge.b)?")
                .font(Theme.body(22, .black))
                .foregroundStyle(Theme.textPrimary)
                .padding(.bottom, 10)
                .accessibilityIdentifier("gate-question")

            TextField("", text: $entry)
                .keyboardType(.numberPad)
                .multilineTextAlignment(.center)
                .font(Theme.body(20, .black))
                // A `TextField`'s text takes the accent colour on this background
                // unless it is told otherwise, which reads as system blue.
                .foregroundStyle(Theme.textPrimary)
                .focused($isFocused)
                .frame(height: Self.fieldHeight)
                .background(Theme.surface, in: RoundedRectangle(cornerRadius: 12, style: .continuous))
                .overlay(
                    RoundedRectangle(cornerRadius: 12, style: .continuous)
                        .stroke(wasWrong ? Theme.coral.opacity(0.6) : Theme.surfaceBorderStrong,
                                lineWidth: 2)
                )
                .onChange(of: entry) { _, new in
                    // A hardware or paste path can still deliver letters.
                    let digits = new.filter(\.isNumber)
                    if digits != new { entry = digits }
                    // Only a **non-empty** edit clears the error, and that is
                    // load-bearing rather than a nicety. React fires `onChange`
                    // for user input only; SwiftUI fires it for any assignment,
                    // including `submit`'s own `entry = ""` — so clearing
                    // unconditionally here sets `wasWrong` back to false in the
                    // same update that set it to true, and the "Not quite" line
                    // never appears at all. Leaving it up while the field is
                    // empty is also the better behaviour: the message stands
                    // until the parent starts a new answer.
                    if !new.isEmpty { wasWrong = false }
                }
                .accessibilityIdentifier("gate-input")
                .padding(.bottom, wasWrong ? 6 : 14)

            if wasWrong {
                Text("Not quite — have another go.")
                    .font(Theme.body(12, .heavy))
                    .foregroundStyle(Theme.coral)
                    .padding(.bottom, 10)
                    .accessibilityIdentifier("gate-error")
            }

            HStack(spacing: 8) {
                gateButton("Cancel", identifier: "gate-cancel",
                           tint: Theme.textTertiary, filled: false, action: onCancel)
                gateButton("Continue", identifier: "gate-submit",
                           tint: Theme.teal, filled: true, action: submit)
            }
        }
        .padding(24)
        .background(
            LinearGradient(colors: [Theme.bgMid, Theme.bgEdge],
                           startPoint: .topLeading, endPoint: .bottomTrailing),
            in: RoundedRectangle(cornerRadius: 16, style: .continuous)
        )
        .overlay(
            RoundedRectangle(cornerRadius: 16, style: .continuous)
                .stroke(Theme.surfaceBorderStrong, lineWidth: 1.5)
        )
        // The web's `onClick={(e) => e.stopPropagation()}`. Without it a tap on
        // the card's own padding — the 24pt margin, or the gap between the
        // question and the field — falls through to the dimmer behind it and
        // cancels the gate a parent is halfway through answering. A child view
        // with its own gesture (the buttons, the field) still wins.
        .contentShape(RoundedRectangle(cornerRadius: 16, style: .continuous))
        .onTapGesture { }
    }

    private func gateButton(
        _ title: String, identifier: String, tint: Color, filled: Bool,
        action: @escaping () -> Void
    ) -> some View {
        Button(action: action) {
            Text(title)
                .font(Theme.body(14, .black))
                // Text, so it takes the accent tint unless this says otherwise.
                .foregroundStyle(tint)
                .frame(maxWidth: .infinity)
                .frame(height: Self.buttonHeight)
                .background(filled ? tint.opacity(0.2) : Color.clear,
                            in: RoundedRectangle(cornerRadius: 12, style: .continuous))
                .overlay(
                    RoundedRectangle(cornerRadius: 12, style: .continuous)
                        .stroke(filled ? tint.opacity(0.4) : Theme.surfaceBorderStrong, lineWidth: 2)
                )
                .contentShape(Rectangle())
        }
        .buttonStyle(PressScale(scale: 0.88))
        .accessibilityIdentifier(identifier)
    }

    private func submit() {
        if challenge.accepts(entry) {
            onPass()
        } else {
            wasWrong = true
            entry = ""
        }
    }
}

#Preview("Gate") {
    ZStack {
        Theme.background.ignoresSafeArea()
        ParentalGate(
            challenge: GateChallenge.at(0),
            action: "Settings let you choose which languages and categories Cloudmoji shows.",
            onPass: {}, onCancel: {}
        )
    }
}

File: android/app/src/main/java/app/cloudmoji/android/ui/parents/ParentalGate.kt
package app.cloudmoji.android.ui.parents

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.gestures.detectTapGestures
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.focus.FocusRequester
import androidx.compose.ui.focus.focusRequester
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.cloudmoji.android.ui.common.pressScale
import app.cloudmoji.android.ui.theme.BackgroundEdge
import app.cloudmoji.android.ui.theme.BackgroundMid
import app.cloudmoji.android.ui.theme.Coral
import app.cloudmoji.android.ui.theme.CloudmojiBodyFont
import app.cloudmoji.android.ui.theme.Surface
import app.cloudmoji.android.ui.theme.SurfaceBorder
import app.cloudmoji.android.ui.theme.Teal
import app.cloudmoji.android.ui.theme.TextPrimary
import app.cloudmoji.android.ui.theme.TextSecondary
import app.cloudmoji.android.ui.theme.TextTertiary

/** Every number [ParentalGate] is drawn from. Mirrors iOS `ParentalGate.swift`'s
 * own constants, pt for dp. Parent-only chrome, so the field/buttons clear the
 * 44dp HIG-style floor rather than the 64dp child-facing one. */
private object GateMetrics {
    val cardWidth = 340.dp
    val fieldHeight = 52.dp
    val buttonHeight = 48.dp
    const val pressedScale = 0.88f
}

/**
 * A real gate, not a gesture. Ported from iOS `ParentalGate.swift`.
 *
 * Deliberately boring: no timer, no penalty, no lock-out — a parent who
 * misreads the question just tries again (see [GateAttempt.submit]). Drawn as
 * a full-screen overlay by the caller (see `CloudmojiApp.kt`) rather than a
 * dialog, so it sits over the launcher tiles *and* whatever mini-app was open
 * when the gear was tapped — a gate a child can tap around underneath is not
 * a gate.
 *
 * [challengeIndex] is owned by the caller and only ever advances by one, on
 * every close (pass or cancel) — this composable's own [GateAttempt] state
 * resets fresh whenever [challengeIndex] changes, exactly like iOS's
 * `ParentalGate` view being torn down and recreated each time the sheet
 * reopens over a new `gateIndex`.
 */
@Composable
fun ParentalGate(
    challengeIndex: Int,
    action: String,
    onPass: () -> Unit,
    onCancel: () -> Unit,
    modifier: Modifier = Modifier,
) {
    // Plain `remember`, not `rememberSaveable` — `GateAttempt` is not
    // `Parcelable`/`Serializable`, and the default Saver would crash trying
    // to put it in a Bundle. Losing an in-progress, unsubmitted digit on a
    // rotation is a fair trade: [challengeIndex] itself lives in the caller's
    // `rememberSaveable` `Int`, so a rotation mid-gate resumes on the same
    // question with a blank field rather than losing the gate outright.
    var attempt by remember(challengeIndex) {
        mutableStateOf(GateAttempt(index = challengeIndex))
    }
    val focusRequester = remember { FocusRequester() }

    LaunchedEffect(challengeIndex) {
        focusRequester.requestFocus()
    }

    fun submit() {
        when (val outcome = attempt.submit()) {
            GateOutcome.Passed -> onPass()
            is GateOutcome.Failed -> attempt = outcome.attempt
        }
    }

    Box(
        modifier = modifier
            .fillMaxSize()
            .background(Color.Black.copy(alpha = 0.82f))
            // Tapping outside the card is Cancel — a toddler who reached this
            // screen needs a way out that is not the right answer. A
            // `pointerInput` rather than `.clickable` so this carries no
            // button semantics of its own; the card beneath swallows its own
            // taps the same way (see below), matching iOS's
            // `.onTapGesture(perform: onCancel)` + the card's own
            // `stopPropagation`-equivalent.
            .pointerInput(onCancel) { detectTapGestures(onTap = { onCancel() }) }
            .semantics { contentDescription = "Grown-ups only" }
            .testTag("parental-gate"),
        contentAlignment = Alignment.Center,
    ) {
        Column(
            modifier = Modifier
                .widthIn(max = GateMetrics.cardWidth)
                .padding(20.dp)
                .clip(RoundedCornerShape(16.dp))
                .background(
                    Brush.linearGradient(listOf(BackgroundMid, BackgroundEdge)),
                    RoundedCornerShape(16.dp),
                )
                .border(1.5.dp, SurfaceBorder, RoundedCornerShape(16.dp))
                .pointerInput(Unit) { detectTapGestures(onTap = {}) }
                .padding(24.dp),
        ) {
            Text(
                text = "Grown-ups only",
                color = TextPrimary,
                fontFamily = CloudmojiBodyFont,
                fontWeight = FontWeight.Black,
                fontSize = 16.sp,
            )
            Spacer(Modifier.height(6.dp))

            Text(
                text = action,
                color = TextSecondary,
                fontFamily = CloudmojiBodyFont,
                fontWeight = FontWeight.Bold,
                fontSize = 13.sp,
            )
            Spacer(Modifier.height(16.dp))

            Text(
                text = "What is ${attempt.challenge.a} × ${attempt.challenge.b}?",
                color = TextPrimary,
                fontFamily = CloudmojiBodyFont,
                fontWeight = FontWeight.Black,
                fontSize = 22.sp,
                modifier = Modifier.testTag("gate-question"),
            )
            Spacer(Modifier.height(10.dp))

            val fieldBorderColor = if (attempt.wasWrong) Coral.copy(alpha = 0.6f) else SurfaceBorder
            Box(
                modifier = Modifier
                    .fillMaxWidth()
                    .height(GateMetrics.fieldHeight)
                    .clip(RoundedCornerShape(12.dp))
                    .background(Surface, RoundedCornerShape(12.dp))
                    .border(2.dp, fieldBorderColor, RoundedCornerShape(12.dp)),
                contentAlignment = Alignment.Center,
            ) {
                BasicTextField(
                    value = attempt.entry,
                    onValueChange = { attempt = attempt.withEntry(it) },
                    singleLine = true,
                    textStyle = TextStyle(
                        color = TextPrimary,
                        fontFamily = CloudmojiBodyFont,
                        fontWeight = FontWeight.Black,
                        fontSize = 20.sp,
                        textAlign = TextAlign.Center,
                    ),
                    cursorBrush = SolidColor(Teal),
                    keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Number),
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(horizontal = 12.dp)
                        .focusRequester(focusRequester)
                        .testTag("gate-input"),
                )
            }
            Spacer(Modifier.height(if (attempt.wasWrong) 6.dp else 14.dp))

            if (attempt.wasWrong) {
                Text(
                    text = "Not quite — have another go.",
                    color = Coral,
                    fontFamily = CloudmojiBodyFont,
                    fontWeight = FontWeight.Black,
                    fontSize = 12.sp,
                    modifier = Modifier.testTag("gate-error"),
                )
                Spacer(Modifier.height(10.dp))
            }

            Row {
                GateButton(
                    title = "Cancel",
                    testTag = "gate-cancel",
                    tint = TextTertiary,
                    filled = false,
                    onClick = onCancel,
                    modifier = Modifier.weight(1f),
                )
                Spacer(Modifier.width(8.dp))
                GateButton(
                    title = "Continue",
                    testTag = "gate-submit",
                    tint = Teal,
                    filled = true,
                    onClick = ::submit,
                    modifier = Modifier.weight(1f),
                )
            }
        }
    }
}

@Composable
private fun GateButton(
    title: String,
    testTag: String,
    tint: Color,
    filled: Boolean,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val interactionSource = remember { MutableInteractionSource() }
    val shape = RoundedCornerShape(12.dp)

    Box(
        contentAlignment = Alignment.Center,
        modifier = modifier
            .height(GateMetrics.buttonHeight)
            .pressScale(interactionSource, GateMetrics.pressedScale)
            .clip(shape)
            .background(if (filled) tint.copy(alpha = 0.2f) else Color.Transparent, shape)
            .border(2.dp, if (filled) tint.copy(alpha = 0.4f) else SurfaceBorder, shape)
            .clickable(
                interactionSource = interactionSource,
                indication = null,
                role = Role.Button,
                onClick = onClick,
            )
            .semantics { role = Role.Button; contentDescription = title }
            .testTag(testTag),
    ) {
        Text(
            text = title,
            color = tint,
            fontFamily = CloudmojiBodyFont,
            fontWeight = FontWeight.Black,
            fontSize = 14.sp,
        )
    }
}
